### PR TITLE
feat: agentOf method

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/types/index.ts [hunk 1]
```diff
@@ -1105,6 +1105,11 @@ export interface GroupedInstructions {
   failed: Instruction[];
 }
 
+export interface AssetPermission {
```
Comment (monitz87) [inline src/types/index.ts:1108]: TokenWithGroup

File: src/api/entities/Identity/index.ts [hunk 2]
```diff
@@ -891,6 +893,37 @@ export class Identity extends Entity<UniqueIdentifiers, string> {
     return assembleResult(didRecords);
   }
 
+  /**
+   * Retrieve all assets with given permissions where this Identity is an Agent
```
Comment (monitz87) [inline src/api/entities/Identity/index.ts:897]: Retrieve all the Security Tokens for which this Identity is an Agent, with the corresponding permission group

File: src/api/entities/Identity/__tests__/index.ts [hunk 3]
```diff
@@ -1521,6 +1530,62 @@ describe('Identity class', () => {
...
+          transactionGroups: [],
+        },
+      });
+
+      entityMockUtils.configureMocks({
+        securityTokenOptions: {
+          ticker,
+          permissionsGetAgents: [
+            {
+              // TODO replace by entityMockUtils.getAgentInstance once it is merge
```
Comment (monitz87) [inline src/api/entities/Identity/__tests__/index.ts:1570]: Why not base this branch off the other?

File: src/types/index.ts [hunk 4]
```diff
@@ -1105,6 +1105,11 @@ export interface GroupedInstructions {
...
+  token: SecurityToken;
+  permissions: GroupPermissions;
```
Comment (monitz87) [inline src/types/index.ts:1110]: I would have this just be `group: KnownPermissionGroup | CustomPermissionGroup`

File: src/api/entities/Identity/index.ts [hunk 5]
```diff
@@ -891,6 +893,37 @@ export class Identity extends Entity<UniqueIdentifiers, string> {
...
+        polymeshApi: {
+          query: { externalAgents },
+        },
+      },
+      did,
+      context,
+    } = this;
+
+    const rawDid = stringToIdentityId(did, context);
+    const rawAgentOf = await externalAgents.agentOf.entries(rawDid);
```
Comment (monitz87) [inline src/api/entities/Identity/index.ts:911]: tokenEntries

File: src/api/entities/Identity/index.ts [hunk 6]
```diff
@@ -891,6 +893,37 @@ export class Identity extends Entity<UniqueIdentifiers, string> {
...
+      const token = new SecurityToken({ ticker: tickerToString(key.args[1]) }, context);
+      const agents = await token.permissions.getAgents();
+      const agentWithGroup = agents.find(({ agent }) => agent.did === did);
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const permissions = await agentWithGroup!.group.getPermissions();
+
+      return {
+        token,
+        permissions,
+      };
```
Comment (monitz87) [inline src/api/entities/Identity/index.ts:923]: const ticker = tickerToString(key.args[1]);
      const agent = new Agent({ did, ticker }, context);
      const token = new SecurityToken({ ticker }, context);
      const group = await agent.getPermissionGroup();

      return {
        token,
        group,
      };

**Review by monitz87 — CHANGES_REQUESTED** [review]

**Comment by monitz87** [comment]
Missing coverage

File: src/api/entities/Identity/__tests__/index.ts [hunk 7]
```diff
@@ -1567,7 +1568,6 @@ describe('Identity class', () => {
           ticker,
           permissionsGetAgents: [
             {
-              // TODO replace by entityMockUtils.getAgentInstance once it is merge
               agent: identity as Agent,
```
Comment (monitz87) [inline src/api/entities/Identity/__tests__/index.ts]: You didn't replace it

File: src/api/entities/Identity/index.ts [hunk 8]
```diff
@@ -894,9 +894,9 @@ export class Identity extends Entity<UniqueIdentifiers, string> {
   }
 
   /**
-   * Retrieve all assets with given permissions where this Identity is an Agent
+   * Retrieve all the Security Tokens for which this Identity is an Agent, with the corresponding permission group
    */
-  public async agentOf(): Promise<AssetPermission[]> {
+  public async agentOf(): Promise<TokenWithGroup[]> {
```
Comment (monitz87) [inline src/api/entities/Identity/index.ts:900]: I think `getTokenPermissions` might be a better fit

File: src/api/entities/Identity/index.ts [hunk 9]
```diff
@@ -908,18 +908,18 @@ export class Identity extends Entity<UniqueIdentifiers, string> {
...
+    const tokenEntries = await externalAgents.agentOf.entries(rawDid);
 
-    return P.map(rawAgentOf, async ([key]) => {
+    return P.map(tokenEntries, async ([key]) => {
       const token = new SecurityToken({ ticker: tickerToString(key.args[1]) }, context);
       const agents = await token.permissions.getAgents();
       const agentWithGroup = agents.find(({ agent }) => agent.did === did);
       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-      const permissions = await agentWithGroup!.group.getPermissions();
+      const group = await agentWithGroup!.group;
```
Comment (monitz87) [inline src/api/entities/Identity/index.ts]: Can't you do

```
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const { group } = agents.find(({ agent }) => agent.did === did)!;
```

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/entities/Identity/__tests__/index.ts (hunk 7 above)
Comment (monitz87) [reply src/api/entities/Identity/__tests__/index.ts]: Why did you mark this as resolved?

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/entities/Identity/__tests__/index.ts (hunk 7 above)
Comment (shuffledex) [reply src/api/entities/Identity/__tests__/index.ts]: 'cos I don't see it anymore

Comment (shuffledex) [reply src/api/entities/Identity/__tests__/index.ts]: the new code is

```
      entityMockUtils.configureMocks({
        securityTokenOptions: {
          ticker,
        },
        agentOptions: {
          getPermissionGroup: group,
        },
      });
```

Comment (monitz87) [reply src/api/entities/Identity/__tests__/index.ts]: All good then

**Review by monitz87 — APPROVED** [review]

**Pull request merged**